File: components/support/nimbus-fml/src/backends/kotlin/templates/ObjectTemplate.kt
{%- import "macros.kt" as kt %}
{%- let inner = self.inner() %}
{%- let class_name = inner.name()|class_name %}

{{ inner.doc()|comment("") }}
public class {{ class_name }} {% call kt::render_constructor() %} : FMLObjectInterface {
    {% call kt::render_class_body(inner) %}

    internal fun _mergeWith(defaults: {{class_name}}?): {{class_name}} =
        defaults?.let { {{class_name}}(_variables = this._variables, _defaults = it._defaults) } ?: this

    companion object {
        internal fun create(variables: Variables): {{class_name}}? =
            {{class_name}}(_variables = variables)

        internal fun mergeWith(overrides: {{class_name}}, defaults: {{class_name}}): {{class_name}} =
            overrides._mergeWith(defaults)
    }
}

File: components/support/nimbus-fml/src/backends/swift/templates/ObjectTemplate.swift
{%- import "macros.swift" as swift %}
{%- let inner = self.inner() %}
{% call swift::render_class(inner) -%}
{%- let class_name = inner.name()|class_name %}

public extension {{class_name}} {
    func _mergeWith(_ defaults: {{class_name}}?) -> {{class_name}} {
        guard let defaults = defaults else {
            return self
        }
        return {{class_name}}(variables: self._variables, prefs: self._prefs, defaults: defaults._defaults)
    }

    static func create(_ variables: Variables?) -> {{class_name}} {
        return {{class_name}}(variables ?? NilVariables.instance)
    }

    static func mergeWith(_ overrides: {{class_name}}, _ defaults: {{class_name}}) -> {{class_name}} {
        return overrides._mergeWith(defaults)
    }
}
